promql/engine.go

```diff
@@ -1345,6 +1346,23 @@ func (ev *evaluator) rangeEvalAgg(aggExpr *parser.AggregateExpr, sortedGrouping
 			return nil, warnings
 		}
 		seriess = make(map[uint64]Series, len(inputMatrix)) // Output series by series hash.
+	case parser.LIMIT_RATIO:
+		if math.IsNaN(param) {
+			ev.errorf("Ratio value %v is NaN", param)
+		}
+		switch {
+		case param == 0:
+			return nil, warnings
+		case param < -1.0:
+			ratio = -1.0
+			warnings.Add(annotations.NewInvalidRatioWarning(param, ratio, aggExpr.Param.PositionRange()))
```

Comment:
Hmm, I just tested this case, and there's no warning being returned from the API 🤔 I don't know why yet though, the `NewInvalidQuantileWarning` warning just below for `quantile()` is working just fine.

Reply:
I've verified with Printfs that it's correctly entering this branch though, so now I wonder even more why the warning doesn't work...

Reply:
Ah, it's because further below the case that handles the limiting operators does a return with a different subset of warnings without first merging in the other set:

https://github.com/prometheus/prometheus/blob/3d54bcc018df3e081758db3e13a49f9a2566c49b/promql/engine.go#L1365-L1370

I guess that return should be changed to:

```go
if ev.endTimestamp == ev.startTimestamp {
	warnings.Merge(ws)
	return result, warnings
}
```

Reply:
oh cool, thanks for thoroughly looking into it!, indeed a couple lines below there's such `Merge(ws)` following the flow -- added it at 883d080c3.